File: flagship-rn-sdk/ios/ios-native/FlagshipSdk.swift
import Foundation
import OpenFeature
import FlagshipFeatureFlags
import os.log

@objc public final class FlagshipSdk: NSObject {
    @objc public static let shared = FlagshipSdk()
    
    private let logger = Logger(subsystem: "com.flagshiprnsdk", category: "FlagshipSdk")
    
    private override init() {
        super.init()
    }
    
    @objc @discardableResult
    public func initialize(
        baseUrl: String,
        apiKey: String,
        refreshInterval: TimeInterval = 10.0
    ) -> Bool {
        if FlagshipState.shared.isInitialized {
            return true
        }
        
        do {
            let config = FlagshipFeatureConfig(
                baseURL: baseUrl,
                refreshInterval: refreshInterval,
                flagshipApiKey: apiKey
            )
            
            let provider = FlagshipOpenFeatureProvider(config: config)
            OpenFeatureAPI.shared.setProvider(provider: provider)
            
            FlagshipState.shared.markInitialized()
            logger.info("Initialized with baseUrl=\(baseUrl), refreshInterval=\(refreshInterval)s")
            return true
        } catch {
            logger.error("Initialization failed: \(error.localizedDescription)")
            return false
        }
    }
    
    @objc @discardableResult
    public func setContext(targetingKey: String, context: [String: Any] = [:]) -> Bool {
        do {
            var attributes: [String: Value] = [:]
            
            for (key, value) in context {
                attributes[key] = convertAnyToValue(value)
            }
            
            let evaluationContext = MutableContext(
                targetingKey: targetingKey,
                structure: MutableStructure(attributes: attributes)
            )
            
            OpenFeatureAPI.shared.setEvaluationContext(evaluationContext: evaluationContext)
            logger.debug("Context set for targetingKey=\(targetingKey)")
            return true
        } catch {
            logger.error("Failed to set context: \(error.localizedDescription)")
            return false
        }
    }
    
    @objc public func getBooleanValue(key: String, defaultValue: Bool) -> Bool {
        do {
            let client = OpenFeatureAPI.shared.getClient()
            return client.getBooleanValue(key: key, defaultValue: defaultValue)
        } catch {
            logger.error("Failed to get boolean value for key=\(key): \(error.localizedDescription)")
            return defaultValue
        }
    }
    
    @objc public func getStringValue(key: String, defaultValue: String) -> String {
        do {
            let client = OpenFeatureAPI.shared.getClient()
            return client.getStringValue(key: key, defaultValue: defaultValue)
        } catch {
            logger.error("Failed to get string value for key=\(key): \(error.localizedDescription)")
            return defaultValue
        }
    }
    
    @objc public func getIntegerValue(key: String, defaultValue: Int) -> Int {
        do {
            let client = OpenFeatureAPI.shared.getClient()
            return Int(client.getIntegerValue(key: key, defaultValue: Int64(defaultValue)))
        } catch {
            logger.error("Failed to get integer value for key=\(key): \(error.localizedDescription)")
            return defaultValue
        }
    }
    
    @objc public func getDoubleValue(key: String, defaultValue: Double) -> Double {
        do {
            let client = OpenFeatureAPI.shared.getClient()
            return client.getDoubleValue(key: key, defaultValue: defaultValue)
        } catch {
            logger.error("Failed to get double value for key=\(key): \(error.localizedDescription)")
            return defaultValue
        }
    }
    
    @objc public func getObjectValue(key: String, defaultValue: [String: Any] = [:]) -> [String: Any] {
        do {
            let client = OpenFeatureAPI.shared.getClient()
            let defaultOpenFeatureValue = convertDictionaryToValue(defaultValue)
            let result = client.getObjectValue(key: key, defaultValue: defaultOpenFeatureValue)
            return convertValueToDictionary(result)
        } catch {
            logger.error("Failed to get object value for key=\(key): \(error.localizedDescription)")
            return defaultValue
        }
    }
    
    @objc public func isInitialized() -> Bool {
        return FlagshipState.shared.isInitialized
    }
    
    private func convertAnyToValue(_ value: Any) -> Value {
        switch value {
        case let boolValue as Bool:
            return .boolean(boolValue)
        case let intValue as Int:
            return .integer(Int64(intValue))
        case let int64Value as Int64:
            return .integer(int64Value)
        case let doubleValue as Double:
            return .double(doubleValue)
        case let floatValue as Float:
            return .double(Double(floatValue))
        case let stringValue as String:
            return .string(stringValue)
        case let dictValue as [String: Any]:
            return convertDictionaryToValue(dictValue)
        case let arrayValue as [Any]:
            return .list(arrayValue.map { convertAnyToValue($0) })
        default:
            return .string(String(describing: value))
        }
    }
    
    private func convertDictionaryToValue(_ dict: [String: Any]) -> Value {
        var result: [String: Value] = [:]
        for (key, value) in dict {
            result[key] = convertAnyToValue(value)
        }
        return .structure(result)
    }
    
    private func convertValueToDictionary(_ value: Value) -> [String: Any] {
        switch value {
        case .structure(let dict):
            var result: [String: Any] = [:]
            for (key, v) in dict {
                result[key] = convertValueToAny(v)
            }
            return result
        default:
            return ["value": convertValueToAny(value)]
        }
    }
    
    private func convertValueToAny(_ value: Value) -> Any {
        switch value {
        case .boolean(let v): return v
        case .integer(let v): return v
        case .double(let v): return v
        case .string(let v): return v
        case .list(let arr): return arr.map { convertValueToAny($0) }
        case .structure(let dict):
            var result: [String: Any] = [:]
            for (k, v) in dict {
                result[k] = convertValueToAny(v)
            }
            return result
        default:
            return String(describing: value)
        }
    }
}

File: flagship-rn-sdk/android/src/main/java/com/flagshiprnsdk/android/FlagshipSdk.kt
package com.flagshiprnsdk.android

import android.app.Application
import android.content.Context
import android.util.Log
import com.flagship.sdk.facade.FlagShipConfig
import com.flagship.sdk.facade.FlagshipProvider
import com.flagshiprnsdk.core.FlagshipState
import dev.openfeature.kotlin.sdk.ImmutableContext
import dev.openfeature.kotlin.sdk.OpenFeatureAPI
import dev.openfeature.kotlin.sdk.Value

object FlagshipSdk {
    private const val TAG = "FlagshipSdk"

    fun initialize(
        context: Context,
        baseUrl: String,
        flagshipApiKey: String,
        refreshIntervalSeconds: Long = 10L
    ): Boolean {
        if (FlagshipState.isInitialized()) {
            return true
        }

        return try {
            val application = context.applicationContext as? Application
                ?: throw IllegalStateException("Unable to get Application context")

            val refreshIntervalMs = refreshIntervalSeconds * 1000

            val flagShipConfig = FlagShipConfig(
                applicationContext = application,
                baseUrl = baseUrl,
                flagshipApiKey = flagshipApiKey,
                refreshInterval = refreshIntervalMs
            )

            val provider = FlagshipProvider("native", flagShipConfig)
            OpenFeatureAPI.setProvider(provider)

            FlagshipState.markInitialized()
            Log.i(TAG, "Initialized with baseUrl=$baseUrl, refreshInterval=${refreshIntervalSeconds}s")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Initialization failed: ${e.message}", e)
            false
        }
    }

    fun setContext(targetingKey: String, context: Map<String, Any?> = emptyMap()): Boolean {
        return try {
            val contextAttributes = mutableMapOf<String, Value>()

            context.forEach { (key, value) ->
                val convertedValue = convertAnyToValue(value)
                contextAttributes[key] = convertedValue
            }

            OpenFeatureAPI.setEvaluationContext(
                ImmutableContext(
                    targetingKey = targetingKey,
                    contextAttributes
                )
            )

            Log.d(TAG, "Context set for targetingKey=$targetingKey")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to set context: ${e.message}", e)
            false
        }
    }

    fun getBooleanValue(key: String, defaultValue: Boolean): Boolean {
        return try {
            val client = OpenFeatureAPI.getClient()
            client.getBooleanValue(key, defaultValue)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get boolean value for key=$key: ${e.message}", e)
            defaultValue
        }
    }

    fun getStringValue(key: String, defaultValue: String): String {
        return try {
            val client = OpenFeatureAPI.getClient()
            client.getStringValue(key, defaultValue)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get string value for key=$key: ${e.message}", e)
            defaultValue
        }
    }

    fun getIntegerValue(key: String, defaultValue: Int): Int {
        return try {
            val client = OpenFeatureAPI.getClient()
            client.getIntegerValue(key, defaultValue)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get integer value for key=$key: ${e.message}", e)
            defaultValue
        }
    }

    fun getDoubleValue(key: String, defaultValue: Double): Double {
        return try {
            val client = OpenFeatureAPI.getClient()
            client.getDoubleValue(key, defaultValue)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get double value for key=$key: ${e.message}", e)
            defaultValue
        }
    }

    fun getObjectValue(key: String, defaultValue: Map<String, Any?> = emptyMap()): Map<String, Any?> {
        return try {
            val client = OpenFeatureAPI.getClient()
            val defaultOpenFeatureValue = convertMapToValue(defaultValue)
            val result = client.getObjectValue(key, defaultOpenFeatureValue)
            convertValueToMap(result)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get object value for key=$key: ${e.message}", e)
            defaultValue
        }
    }

    fun isInitialized(): Boolean = FlagshipState.isInitialized()

    private fun convertAnyToValue(value: Any?): Value {
        return when (value) {
            null -> Value.Null
            is Boolean -> Value.Boolean(value)
            is Int -> Value.Integer(value)
            is Long -> Value.Integer(value.toInt())
            is Double -> Value.Double(value)
            is Float -> Value.Double(value.toDouble())
            is String -> Value.String(value)
            is Map<*, *> -> {
                @Suppress("UNCHECKED_CAST")
                convertMapToValue(value as Map<String, Any?>)
            }
            is List<*> -> {
                val list = value.map { convertAnyToValue(it) }
                Value.List(list)
            }
            else -> Value.String(value.toString())
        }
    }

    private fun convertMapToValue(map: Map<String, Any?>): Value {
        val result = mutableMapOf<String, Value>()
        map.forEach { (key, value) ->
            result[key] = convertAnyToValue(value)
        }
        return Value.Structure(result)
    }

    private fun convertValueToMap(value: Value): Map<String, Any?> {
        return when (value) {
            is Value.Structure -> {
                val result = mutableMapOf<String, Any?>()
                value.structure.forEach { (key, v) ->
                    result[key] = convertValueToAny(v)
                }
                result
            }
            else -> mapOf("value" to convertValueToAny(value))
        }
    }

    private fun convertValueToAny(value: Value): Any? {
        return when (value) {
            is Value.Null -> null
            is Value.Boolean -> value.boolean
            is Value.Integer -> value.integer
            is Value.Double -> value.double
            is Value.String -> value.string
            is Value.List -> value.list.map { convertValueToAny(it) }
            is Value.Structure -> {
                val result = mutableMapOf<String, Any?>()
                value.structure.forEach { (k, v) ->
                    result[k] = convertValueToAny(v)
                }
                result
            }
            else -> value.toString()
        }
    }
}

